### PR TITLE
post: async await refactor

### PR DESCRIPTION
**Pull request by alanshaw**
Two blog posts: one that covers the reasons and benefits of the async/await refactor and one that covers some of the learnings of completing a big code refactor with a distributed team.

### CODE REVIEW

File: content/post/076-async-await-refactor.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+We're on the cusp of completing a refactor to the js-ipfs, js-libp2p and js-ipld codebases to use Promises and remove Node.js/pull streams from the code base entirely. We're using `async`/`await` everywhere (i.e. not the `then`/`catch` style of working with promises) and async iterables, allowing users to consume our streaming APIs with `for await (const chunk of iterable)` loops.
+
+First up, if you think you'd like to take on a challenge like this, do the math. If each repo takes a day to refactor and if you work on it 5 days a week then you've automatically used 14 of your 52 weeks in a year. That's like, 4 months, and that's the absolute best case scenario.
+
+In reality, even if you get approval to do it you're unlikely to be able to work on it all day every day. The problem is that the benefits of a refactor like this can't really be seen until everything is done and the whole stack is using the new code. I'll go into the benefits more later, but the point is, **your project needs to keep moving**. New features need to land and those bugs don't squash themselves.
```
Comment (momack2) [inline content/post/076-async-await-refactor.md]: You say you’ll get into benefits later but don’t here. Edgy about actually starting with this as a “whew, we got to this amazing summit and now we’re going to reflect back on the climb”? I don’t feel like I get a sense of why all this work was relevant or worthwhile in the current structure - I need convincing first that the destination is even interesting/exciting before I’m ready to know how you achieved this long, hard feat...

File: content/post/076-async-await-refactor.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+Refactors to each dependent repo **won't take just one day**. A tiny percentage of them will take less than a day but the vast majority will take multiple days to complete, if not weeks.
```
Comment (momack2) [inline content/post/076-async-await-refactor.md]: These seem like some good general lessons for hard ref actors. Maybe give this section a subtitle and associated intro?

File: content/post/079-async-await-refactor.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+**Removing streaming alternatives reduces our API surface area**. That's good because it means there's simply fewer API methods for developers to understand and choose from. This just makes things a little easier, and by guiding developers towards streaming by default we're hopefully helping them to avoid OOM errors when their applications become popular or are used on resource constrained devices like mobile or when their users start sharing huge files.
+
+A more concrete benefit of a smaller API surface area is that there is simply less code to ship. There's a lot less boilerplate and many fewer streaming modules and conversion utilities that need to be `npm install`-ed or that make their way into the final bundle. That means a smaller `node_modules` directory and smaller browser builds for applications. There's also no performance hits to be had converting between two different streaming implementations.
```
Comment (achingbrain) [inline content/post/079-async-await-refactor.md]: Maybe we could mention the LOC count before and after the refactor as an example?

File: content/post/080-big-refactors.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+In reality, even if you get approval to do it you're unlikely to be able to work on it all day every day. The problem is that the benefits of a refactor can't always be seen until everything is done and the whole stack is using the new code. The point is, **your project needs to keep moving**. New features need to land and those bugs don't squash themselves.
+
+Refactors to each dependent repo **won't take just one day**. A tiny percentage of them will take less than a day but the vast majority will take multiple days to complete, if not weeks.
+
+You'll be tempted to make those API changes and performance refactors you've always wanted to make. **Beware of scope creep**. If you also decide to take on these tasks, be aware that you're adding to the complexity of integrating a massive breaking change and also to the time the refactor is going to take. You'll probably significantly underestimate the amount of extra time your additional changes will take to write, be tested, be reviewed and re-integrated into the project. **Stay focused**. You cannot make everything perfect now. Pick your battles, open an issue, and move on.
```
Comment (achingbrain) [inline content/post/080-big-refactors.md]: You are also likely to introduce new bugs with your 'improvements' and not spend the same sort of time thinking about them because you are distracted by your flow control refactor.

File: content/post/080-big-refactors.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+Some repos will just take days or weeks to actually convert, and that's not even including converting the tests, getting your work reviewed, making changes from review feedback or rebasing your work because master got this critical security fix in the interim.
+
+Prioritise refactors by dependents. It's obvious, but maybe worth highlighting - the only way you're going to get this done is from the bottom up. **Refactor repos that don't depend on anything first**. Make a list and assign an approximate priority to each repo based on how many other repos it depends on.
```
Comment (achingbrain) [inline content/post/080-big-refactors.md]: This creates an opportunity to example your project's structure and cross-module interdependencies.  We removed a lot of dev-deps on IPFS and libp2p from modules low in the stack as part of this refactor, speeding up test runs and making CI significantly more reliable.

File: content/post/079-async-await-refactor.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+## 4. Improved readability
+
+* easier to follow code
+* I prefer callbacks to then/catch but async/await is so much easier to follow
+
+## 5. Reduced boilerplate
+## 6. Reduced bundle size
+## 7. More performant
+
+* Finally, we still need to verify, but connection setup and data transfer may be faster now. In benchmarks taken at the mplex and lower levels we already know that performance has improved. One aspect of this is likely to be `BufferList`.
```
Comment (achingbrain) [inline content/post/079-async-await-refactor.md]: Use some of the stats from the before/after interop test speeds here?

File: content/post/079-async-await-refactor.md (hunk 3 above)
Comment (achingbrain) [reply content/post/079-async-await-refactor.md]: Also the big size drop for `ipfs-http-client`, presumably we'll see something similar when `ipfs` gets an rc

File: content/post/079-async-await-refactor.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+🚨 If you're using **js-ipfs** or **js-ipfs-http-client**, this is your early warning klaxon - the next releases of these two modules will have BIG breaking changes. Good news though, the [migration guide](https://gist.github.com/alanshaw/04b2ddc35a6fff25c040c011ac6acf26) is already written and a pre-release of js-ipfs-http-client is already available for testing (`v42.0.0-pre.0`) along with a comprehensive [list of breaking changes](https://github.com/ipfs/js-ipfs-http-client/releases/tag/v42.0.0-pre.0). So, read this post and then check out those resources, it'll make the transition way easier, I promise you (no pun intended).
```
Comment (achingbrain) [inline content/post/079-async-await-refactor.md]: I can't add line comments to the migration guide, but it mentions creating PeerIds from CIDs, will need to change to creating them from strings instead.

File: content/post/079-async-await-refactor.md (hunk 3 above)
Comment (alanshaw) [reply content/post/079-async-await-refactor.md]: I was saving the perf wins for the release blog post...

File: content/post/079-async-await-refactor.md (hunk 7 above)
Comment (alanshaw) [reply content/post/079-async-await-refactor.md]: 👍 thanks - updated

File: content/post/079-async-await-refactor.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+When the size of the result is small and/or the information is readily available it is absolutely fine to have, and use, buffering APIs; we're not suggesting switching _all_ API methods to be streaming by default. However, when the size of the result is big or unknown and/or may not be instantly available it makes a whole lot more sense for it to be streamed.
+
+**For big files, it's entirely necessary to stream data to reduce memory pressure**. This is paricularly important for IPFS because typically users will not know the size of a file, given a hash, and it will be obtained from a network of poorly connected peers where locality and immediacy cannot be assumed. Bearing that in mind, we should provide APIs that a) encourage developers to do the right thing and not buffer content into memory regardless of the size of the file they're retrieving and b) reduce time to first byte, giving users visibility over progress and allows them to differentiate between the file being unavailable or it being very large.
```
Comment (jimpick) [inline content/post/079-async-await-refactor.md]: s/paricularly/particularly/

File: content/post/079-async-await-refactor.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+Likewise, if you're writing a function that takes a callback you've got to be _really_ careful that it doesn't throw, or make sure you catch any potential throws and pass them to the callback.
+
+With promises the situation is significantly better, but you can still be in danger of an unhandled rejection if code in a `.catch()` throws.
+
+With `async`/`await` you use only `try`/`catch`, and it works for all errors: synchronous and asynchronous. It reduces boilerplate and guards against uncaught exceptions and unhandled rejections making it significanly less likely your code will take down the process or unrecoverably crash your application.
+
+The other super power that `try`/`catch` has is in combination with [`for await...of`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of) loops to stream data. I find it amazing that you can just `catch` an error in a stream and continue program execution. Let me explain why:
```
Comment (jimpick) [inline content/post/079-async-await-refactor.md]: I find the first sentence here hard to parse...

File: content/post/079-async-await-refactor.md [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+If you're familiar with Node.js streams then you'll know that you have to add an event listener for the `error` event, somehow deal with the error and continue program execution. You also need to listen for the `data` event _or_ "pipe" the stream somewhere and listen for an event that signifies that the stream has completed. Writable streams have `close` and `finish` events and you have to figure out the correct one to use, while worrying about whether the `close` listener gets called or not when an `error` occurs. Also, readable streams emit an `end` event, so, don't get confused between `end`, `close` and `finish`...or do, because y'all are humans.
```
Comment (jimpick) [inline content/post/079-async-await-refactor.md]: perhaps change "somehow deal with the error" to "and then somehow deal with the error" ?

Comment (jacobheun) [inline content/post/079-async-await-refactor.md]: Just a language suggestion to change the reading from "you'll have a hard time keeping this straight" to "it's hard to keep straight for everyone" 
```suggestion
If you're familiar with Node.js streams then you'll know that you have to add an event listener for the `error` event, somehow deal with the error and continue program execution. You also need to listen for the `data` event _or_ "pipe" the stream somewhere and listen for an event that signifies that the stream has completed. Writable streams have `close` and `finish` events and you have to figure out the correct one to use, while worrying about whether the `close` listener gets called or not when an `error` occurs. Also, readable streams emit an `end` event, so, don't get confused between `end`, `close` and `finish`...or do, because we're all humans.
```

File: content/post/079-async-await-refactor.md [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+This is also true when working with conditionals. Things get hairy pretty quickly as soon as you need to conditionally execute async tasks and then continue with some common code. There's libraries that'll help with this but just being able to use an `if` statement is so much easier and doesn't require the cognitive overhead of understanding those libraries.
+
+Readability is also improved by the reduction in overall code to look at. By this I mean that there's a lot less boilerplate involved with asychronous code written using `async`/`await`. In refactoring our callback based code we often found ourselves reducing 3 (or more) lines of code down to a single statement that achieved exactly the same task.
+
+## 5. Performance improvements
+
+The reduction in boilerplate in combination with the reduced API surface area (which meant we could remove many modules relating to different stream implementations) means **js-ipfs will be faster to `npm install`, take up less disk space and be smaller when bundled with your dweb applications**. A smaller bundle is great for saving bandwidth, but also faster to download and more friendly in low powered or resource constrained environments.
```
Comment (jacobheun) [inline content/post/079-async-await-refactor.md]:
```suggestion
The reduction in boilerplate in combination with the reduced API surface area (which meant we could remove many modules relating to different stream implementations) means **js-ipfs will be faster to `npm install`, take up less disk space and be smaller when bundled with your dweb applications**. A smaller bundle is great for saving bandwidth, but also faster to download and friendlier in low powered or resource constrained environments.
```

File: content/post/080-big-refactors.md [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+It's not always possible to use a refactored repo in your dependencies. Sometimes this is due to a circular dependency but sometimes the refactor in another repo is taking longer than expected or is blocked on something else. **Unblock yourself with facades**. You can make the old look like the new but consider it carefully. Doing this takes time and will need to be revisited at a later date to be removed once the dependency refactor is released. In our refactor we made use of `promisify` and `callbackify` modules and also created facades to make the new look like the old so that it could get into production sooner.
+
```
Comment (jacobheun) [inline content/post/080-big-refactors.md]: Most of this is pretty generic to any large refactor, but I wonder if it would be valuable to list some npm release learnings. There were a few dependencies we released early that were not backwards compatible and we had a lot of users issues trying to use the new with the old. Leverage `beta` tags in npm allowed us to do these releases earlier, roll them up into their dependents and avoid issues with users pulling down releases early. The beauty about this is also that when it's time to go live, it's a simple switch of dist-tags on npm. No new code needs to get deployed, which also means you're guaranteed to be using the exact code you tested with.

File: content/post/079-async-await-refactor.md [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+
+When the size of the result is small and/or the information is readily available it is absolutely fine to have, and use, buffering APIs; we're not suggesting switching _all_ API methods to be streaming by default. However, when the size of the result is big or unknown and/or may not be instantly available it makes a whole lot more sense for it to be streamed.
+
+**For big files, it's entirely necessary to stream data to reduce memory pressure**. This is particularly important for IPFS because typically users will not know the size of a file, given a hash, and it will be obtained from a network of poorly connected peers where locality and immediacy cannot be assumed. Bearing that in mind, we should provide APIs that a) encourage developers to do the right thing and not buffer content into memory regardless of the size of the file they're retrieving and b) reduce time to first byte, giving users visibility over progress and allows them to differentiate between the file being unavailable or it being very large.
```
Comment (momack2) [inline content/post/079-async-await-refactor.md]:
```suggestion
**For big files, it's entirely necessary to stream data to reduce memory pressure**. This is particularly important for IPFS because typically users will not know the size of a file, given a hash, and it will be obtained from a network of poorly connected peers where locality and immediacy cannot be assumed. Bearing that in mind, we should provide APIs that a) encourage developers to do the right thing and not buffer content into memory regardless of the size of the file they're retrieving and b) reduce time to first byte, giving users visibility over progress and allowing them to differentiate between the file being unavailable or it being very large.
```

File: content/post/079-async-await-refactor.md [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+
+A more concrete benefit of a smaller API surface area is that there is simply less code to ship. There's a lot less boilerplate and many fewer streaming modules and conversion utilities that need to be `npm install`-ed or that make their way into the final bundle. That means a smaller `node_modules` directory and smaller browser builds for applications. There's also no performance hits to be had converting between two different streaming implementations.
+
+## 2. Streaming with async iterables
+
+Our streaming APIs will use [async iterables](https://javascript.info/async-iterators-generators). Thats significant for a couple of reasons. Firstly, since we've been talking about bundle sizes it's worth mentioning that using a streaming implementation that is actually a language feature means that we don't need to _bundle_ any libraries/modules to provide a streaming implementation! Obviously that also means that we don't need to `npm install` them either.
```
Comment (momack2) [inline content/post/079-async-await-refactor.md]:
```suggestion
Our streaming APIs will use [async iterables](https://javascript.info/async-iterators-generators). That's significant for a couple of reasons. Firstly, since we've been talking about bundle sizes it's worth mentioning that using a streaming implementation that is actually a language feature means that we don't need to _bundle_ any libraries/modules to provide a streaming implementation! Obviously, that also means that we don't need to `npm install` them either.
```

File: content/post/079-async-await-refactor.md [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+
+Secondly, **async iterables make streaming APIs approachable**. There's a BIG hurdle involved with streaming APIs that is lowered significantly by using async iterables instead of Node.js or pull streams. The hurdle is the understanding of the concepts, naming and usage that surrounds them that developers new to JS or new to streaming have to overcome.
```
Comment (momack2) [inline content/post/079-async-await-refactor.md]:
```suggestion
Secondly, **async iterables make streaming APIs approachable**. There's a BIG hurdle involved with streaming APIs that is lowered significantly by using async iterables instead of Node.js or pull streams. That hurdle is understanding the naming, usage, and concepts, which developers new to JS or new to streaming have to overcome.
```
Found the previous sentence hard to parse. Is this better?

File: content/post/079-async-await-refactor.md [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+
+There's a big conflict of interests in exposing streaming APIs and exposing an API that is easy to use and understand for _all_ developers. When you boil down the understanding of reading from a stream to a `for` loop (you read an async iterable using [`for await...of`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of) loops) then developers can instantly draw parallels to one of the first things they learnt in programming. This is simply not the case if your API returns a Node.js `ReadableStream` for example.
```
Comment (momack2) [inline content/post/079-async-await-refactor.md]:
```suggestion
There's tension between exposing streaming APIs vs exposing an API that is easy to use and understand for _all_ developers. With async iterables, you can boil down the understanding of reading from a stream to a `for` loop (you read an async iterable using [`for await...of`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of) loops), helping developers instantly draw parallels to one of the first things they learnt in programming and making streaming APIs easier to use. This is simply not the case if your API returns a Node.js `ReadableStream` for example.
```
Like the point, but wanted to spell out the connection more clearly for us non-js folks...better?

File: content/post/079-async-await-refactor.md [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+
+## 3. Better debugging
+
+One of the best things about debugging in an `async`/`await` code base is that **stack traces are much better**. With callbacks you typically lose the stack trace across an async boundary because with callbacks, the call stack starts at the callback function.
+
+This can make debugging super difficult because you can't get a good idea of the execution path of the program prior to the async boundary. The situation is improved by using promises but by using `async`/`await` it's _even better_:
```
Comment (momack2) [inline content/post/079-async-await-refactor.md]:
```suggestion
This can make debugging super difficult because you can't get a good idea of the execution path of the program prior to the async boundary. The situation is improved by using promises, but using `async`/`await` is _even better_:
```

File: content/post/079-async-await-refactor.md [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+
+> `await X()` suspends execution of the current function, while `promise.then(X)` continues execution of the current function after adding the X call to the callback chain. In the context of stack traces, this difference is pretty significant.
+> https://mathiasbynens.be/notes/async-stack-traces
+
+The upshot is that stack traces are more informative and span async boundaries giving developers more power to debug code.
+
+The second best thing about debugging in an `async`/`await` code base is that synchronous and asynchronous errors are handled by the same construct: `try`/`catch`.
+
+Asynchronous errors when using callbacks must be handled by checking for an error parameter passed to the callback function using an `if` statement. You must also either _trust_ that a call to a function that takes a callback does not throw synchronously (this is what most people do) or put a `try`/`catch` around it as well.
```
Comment (momack2) [inline content/post/079-async-await-refactor.md]: "trust that a call doesn't throw synchronously" -- is that a correct saying in js-land or is "trust that a call doesn't throw a synchronous error" more grammatically correct?

Comment (momack2) [reply content/post/079-async-await-refactor.md]: (I see a lot of this "throw" below - so if normal maybe explain what it is the first time so I don't get lost with this new term?)

File: content/post/079-async-await-refactor.md (hunk 9 above)
Comment (momack2) [reply content/post/079-async-await-refactor.md]: agreed - what is the super power?

File: content/post/079-async-await-refactor.md [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+
+The `async`/`await` syntax, async iterables and `for await...of` loops are now supported by the overwhelming majority of browsers as well as Node.js since v10. It means that we can continue to _not_ transpile our code, keeping debugging nice and easy.
+
+Another great reason to make this switch is that **Node.js streams are async iterable**, and have been for a while now. So we can switch to async iterables and remove Node.js streams from the code base, but that doesn't stop us from accepting them as inputs because we don't need any special streaming code to be able to read from them. Browser streams will also hopefully follow this move too.
+
+In rewriting the code base we built a few tools to help us work with async iterables and share code. They mostly begin with the `it-` prefix (for "iterable") which follows other existing module themes like the `p-` ("promises") collection of modules. They're being documented here: https://github.com/alanshaw/it-awesome
+
+What's funny about these tools is that there's a iterator helpers proposal (https://github.com/tc39/proposal-iterator-helpers) at stage 2 of the TC39 process, which means that:
```
Comment (momack2) [inline content/post/079-async-await-refactor.md]: is funny the right word here? Is "exciting" better?

File: content/post/079-async-await-refactor.md (hunk 18 above)
Comment (momack2) [reply content/post/079-async-await-refactor.md]: Reading through - this is the section that really lost me. there's a lot of detail between each comparison and it's hard to know where to skip to to get out of the debugging details. Maybe either try cutting back or combining into one technically dense paragraph so non-js readers can easily skip to the next section if this isn't relevant to them?

File: content/post/080-big-refactors.md [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+Some repos will just take days or weeks to actually convert, and that's not even including converting the tests, getting your work reviewed, making changes from review feedback or rebasing your work because master got this critical security fix in the interim.
+
+Prioritise refactors by dependents. It's obvious, but maybe worth highlighting - the only way you're going to get this done is from the bottom up. **Refactor repos that don't depend on anything first**. Make a list and assign an approximate priority to each repo based on how many other repos it depends on. This creates an opportunity to example your project's structure and cross-module interdependencies. We removed a lot of dev-deps on IPFS and libp2p from modules low in the stack as part of this refactor, speeding up test runs and making CI significantly more reliable.
```
Comment (momack2) [inline content/post/080-big-refactors.md]:
```suggestion
Prioritise refactors by dependents. It's obvious, but maybe worth highlighting - the only way you're going to get this done is from the bottom up. **Refactor repos that don't depend on anything first**. Make a list and assign an approximate priority to each repo based on how many other repos it depends on. This creates an opportunity to audit your project's structure and cross-module interdependencies. We removed a lot of dev-deps on IPFS and libp2p from modules low in the stack as part of this refactor, speeding up test runs and making CI significantly more reliable.
```

File: content/post/080-big-refactors.md [hunk 21]
```diff
@@ -0,0 +1,26 @@
+---
+date: 2020-01-31
```
Comment (momack2) [inline content/post/080-big-refactors.md]: Maybe delay this by a day or two? If you release both at the same time it's more likely folks will read this post *before* the more detailed one above.

File: content/post/080-big-refactors.md [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+url: 2020-01-31-big-refactors
+title: Big Refactors
+author: Alan Shaw
+---
+
+> If you're considering a refactor that'll touch **~70** interdependent repos you've come to the right place for some perspective! This is the sister post to [The Async Await Refactor](/2020-01-28-async-await-refactor/) and covers some of the learnings of completing a big code refactor with a distributed team.
```
Comment (momack2) [inline content/post/080-big-refactors.md]:
```suggestion
> If you're considering a refactor that'll touch **~70** interdependent repos you've come to the right place for some perspective! This is the sister post to [The Async Await Refactor](/2020-01-31-async-await-refactor/) and covers some of the learnings of completing a big code refactor with a distributed team.
```

File: content/post/079-async-await-refactor.md (hunk 18 above)
Comment (alanshaw) [reply content/post/079-async-await-refactor.md]: I think "synchronously" can just be removed from that sentence.

I'll try to fix up...

Programming languages typically have some sort of concept of throwing or raising an error/exception that can be caught. The golang equivalent is panic/recover.

Comment (alanshaw) [reply content/post/079-async-await-refactor.md]: Ok, I've expanded this section considerably with examples and better explanations. Hopefully it's better now 🤞

File: content/post/076-async-await-refactor.md (hunk 1 above)
Comment (alanshaw) [reply content/post/076-async-await-refactor.md]: Split into 2 posts to achieve this :)
